### PR TITLE
add spring security resource server sample

### PR DESCRIPTION
**Pull request by ZhuXiaoBing-cn**
* add spring security resource server sample
* update README file

### CODE REVIEW

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
+# OAuth 2.0 Sample for Azure AD Spring Boot Starter Resource Server library for Java
+
+## Key concepts
+This sample illustrates how to protect an Java web API by restricting access to its resources to authorized accounts only.
```
Comment (backwind1233) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:4]: an Java --> a java?

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+Obtain Bearer Token in the request head of resources and carry out analysis and verification. After verification,
+injects AzureOAuth2AuthenticatedPrincipal object that is associated with the thread of the current user request. 
+Token claimsset, Authorities, JWS headers etc. are accessible from the object which can be used for role based authorization.
+
+
+## Getting started
+
+To run this sample, you'll need:
+
+- A [Java Development Kit (JDK)][jdk_link], version 1.8 and above.
```
Comment (backwind1233) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:14]: version 1.8 and above --> version 1.8 or above

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+- [Maven](https://maven.apache.org/) 3.0 and above
```
Comment (backwind1233) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:15]: 3.0 and above --> 3.0  or above

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
+# OAuth 2.0 Sample for Azure AD Spring Boot Starter Resource Server library for Java
+
+## Key concepts
+This sample illustrates how to protect an Java web API by restricting access to its resources to authorized accounts only.
+Obtain Bearer Token in the request head of resources and carry out analysis and verification. After verification,
+injects AzureOAuth2AuthenticatedPrincipal object that is associated with the thread of the current user request. 
+Token claimsset, Authorities, JWS headers etc. are accessible from the object which can be used for role based authorization.
```
Comment (backwind1233) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:7]: I feel a little confusing by the sentence here at first.

**Review by backwind1233 — COMMENTED** [review]
pls review

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md (hunk 4 above)
Comment (ZhuXiaoBing-cn) [reply sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:7]: I'm going to modify it.

**Review by backwind1233 — COMMENTED** [review]
LGTM

**Review by backwind1233 — APPROVED** [review]
LGTM

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+## Key concepts
+This sample illustrates how to protect a Java web API by restricting access to its resources to authorized accounts only.
+
+1. The bearer token is obtained from the request header. 
+2. `JwtDecoder`  is used to parse the token into `Jwt`.
+3. Claims, Headers etc in `Jwt` will be extracted, they will be wrapped in `AzureOAuth2AuthenticatedPrincipal` object.
+4. `AzureOAuth2AuthenticatedPrincipal` will eventually be set into SecurityContext.
+
+
+## Getting started
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:12]: We refactor this and [instruction] below in @moarychan's PR, please follow this https://github.com/Azure/azure-sdk-for-java/pull/17862 to change.

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+```properties
+azure.activedirectory.app-id-uri=xxxxxxxx-app-id-uri-xxxxxxxxxx
+azure.activedirectory.session-stateless=true
+#Use a port that is not occupied
+server.port=8081
+```
+
+### Run with Maven 
+First, we need to ensure that this [instruction] is completed before run.
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:90]: Need to change this too.

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
+# OAuth 2.0 Sample for Azure AD Spring Boot Starter Resource Server library for Java
+
+## Key concepts
+This sample illustrates how to protect a Java web API by restricting access to its resources to authorized accounts only.
+
+1. The bearer token is obtained from the request header. 
+2. `JwtDecoder`  is used to parse the token into `Jwt`.
+3. Claims, Headers etc in `Jwt` will be extracted, they will be wrapped in `AzureOAuth2AuthenticatedPrincipal` object.
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:8]: Change this `H` in `Headers` to lowercase.

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+1. Go to the Microsoft identity platform for developers App registrations portal.
+
+2. Select New registration.
+    ![Select New registration](docs/image-select-new-registration.png "Select new registration")
+
+3. When the Register an application page opens, enter your application's registration information:
+    ![Scope Config](docs/image-register-an-application.png "Register an application")
+    
+4. In the **Expose an API** section, select **Add a scope**, accept the proposed Application ID URI `(api://{clientId})` (back up the Application ID URI here,which will be used in the properties file) by selecting **Save and Continue**.
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:65]: Please add a space before `which`

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+
+## Examples
+
+### Configure application.properties
+
+```properties
+azure.activedirectory.app-id-uri=xxxxxxxx-app-id-uri-xxxxxxxxxx
+azure.activedirectory.session-stateless=true
+#Use a port that is not occupied
+server.port=8081
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:86]: This configuration entry is not required, right? Perhaps we could remove this.

File: sdk/spring/azure-spring-boot-starter-active-directory/README.md [hunk 10]
```diff
@@ -263,6 +263,52 @@ logging.level.org.hibernate=ERROR
 For more information about setting logging in spring, please refer to the [official doc](https://docs.spring.io/spring-boot/docs/current/reference/html/spring-boot-features.html#boot-features-logging).
  
 
+### Protect the Resource API through Resource Server
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-starter-active-directory/README.md:266]: We should move this under the ## Examples section.

File: sdk/spring/azure-spring-boot-starter-active-directory/README.md [hunk 11]
```diff
@@ -263,6 +263,52 @@ logging.level.org.hibernate=ERROR
...
+Please refer to [azure-spring-boot-sample-active-directory-spring-oauth2-resource-server](https://github.com/Azure/azure-sdk-for-java/blob/master/sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md) for access resource api.
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-starter-active-directory/README.md:267]: Please use reference link style here.

File: sdk/spring/azure-spring-boot-starter-active-directory/README.md [hunk 12]
```diff
@@ -263,6 +263,52 @@ logging.level.org.hibernate=ERROR
...
+    </dependency>
+  </dependencies>
+```
+
+####  Configure application.properties:
+```properties
+azure.activedirectory.app-id-uri=xxxxxxxx-app-id-uri-xxxxxxxxxx
+azure.activedirectory.session-stateless=true
+#Use a port that is not occupied
+server.port=8081
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-starter-active-directory/README.md:292]: not required.

File: sdk/spring/azure-spring-boot-starter-active-directory/README.md [hunk 13]
```diff
@@ -263,6 +263,52 @@ logging.level.org.hibernate=ERROR
...
+      <groupId>org.springframework.security</groupId>
+      <artifactId>spring-security-oauth2-jose</artifactId>
+    </dependency>
+  </dependencies>
+```
+
+####  Configure application.properties:
+```properties
+azure.activedirectory.app-id-uri=xxxxxxxx-app-id-uri-xxxxxxxxxx
+azure.activedirectory.session-stateless=true
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-starter-active-directory/README.md:290]: Why is this needed?

**Review by saragluna — CHANGES_REQUESTED** [review]

**Comment by saragluna** [comment]
https://github.com/Azure/azure-sdk-for-java/issues/18065

File: sdk/spring/azure-spring-boot-starter-active-directory/README.md [hunk 14]
```diff
@@ -284,14 +266,6 @@ Please refer to [azure-spring-boot-sample-active-directory-spring-oauth2-resourc
   </dependencies>
 ```
 
-####  Configure application.properties:
-```properties
-azure.activedirectory.app-id-uri=xxxxxxxx-app-id-uri-xxxxxxxxxx
-azure.activedirectory.session-stateless=true
-#Use a port that is not occupied
-server.port=8081
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-starter-active-directory/README.md]: Sorry for didn't make it clear last time, I mean the configuration `server.port=8081` is not required last time

**Review by saragluna — CHANGES_REQUESTED** [review]
Sorry for the trouble, we could add back the `####  Configure application.properties:` section, without the `server.port` property entry.

File: sdk/spring/azure-spring-boot-starter-active-directory/README.md (hunk 14 above)
Comment (ZhuXiaoBing-cn) [reply sdk/spring/azure-spring-boot-starter-active-directory/README.md]: okay

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+- An OS X machine (necessary if you want to run the app on Mac)
+- A Linux machine (necessary if you want to run the app on Linux)
+- An Azure Active Directory (Azure AD) tenant. For more information on how to get an Azure AD tenant, see [How to get an Azure AD tenant](https://azure.microsoft.com/documentation/articles/active-directory-howto-tenant/)
+- You register your web APP in App registrations in the Azure portal.
+- A Web APP runtime that requires access to a Web API.
+
+### Environment checklist
+We need to ensure that this [environment checklist][ready-to-run-checklist] is completed before the run.
+
+## Include the package
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:29]: test

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 16]
```diff
@@ -0,0 +1,125 @@
+# OAuth 2.0 Sample for Azure AD Spring Boot Starter Resource Server library for Java
+
+## Key concepts
+This sample illustrates how to protect a Java web API by restricting access to its resources to authorized accounts only.
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:4]: test

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+
+1. The bearer token is obtained from the request header. 
+2. `JwtDecoder`  is used to parse the token into `Jwt`.
+3. Claims, headers etc in `Jwt` will be extracted, they will be wrapped in `AzureOAuth2AuthenticatedPrincipal` object.
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:8]: test

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 18]
```diff
@@ -0,0 +1,125 @@
+# OAuth 2.0 Sample for Azure AD Spring Boot Starter Resource Server library for Java
+
+## Key concepts
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:3]: test

**Review by saragluna — COMMENTED** [review]
test

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 19]
```diff
@@ -0,0 +1,125 @@
...
+```properties
+azure.activedirectory.client-id=xxxxxx-your-client-id-xxxxxx
+azure.activedirectory.app-id-uri=xxxxxxxx-app-id-uri-xxxxxxxxxx
+azure.activedirectory.session-stateless=true
+```
+
+### Run with Maven 
+```shell
+# Under sdk/spring project root directory
+mvn clean install -DskipTests
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:96]: This is not needed, in @moarychan's PR we already included the install part.

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 20]
```diff
@@ -0,0 +1,125 @@
...
+## Getting started
+
+To run this sample, you'll need:
+
+- A [Java Development Kit (JDK)][jdk_link], version 1.8 or above.
+- [Maven](https://maven.apache.org/) 3.0 or above
+- An Internet connection
+- A Windows machine (necessary if you want to run the app on Windows)
+- An OS X machine (necessary if you want to run the app on Mac)
+- A Linux machine (necessary if you want to run the app on Linux)
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:21]: These requirements are included in the environment checklist already.

File: sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md [hunk 21]
```diff
@@ -0,0 +1,125 @@
...
+- An Azure Active Directory (Azure AD) tenant. For more information on how to get an Azure AD tenant, see [How to get an Azure AD tenant](https://azure.microsoft.com/documentation/articles/active-directory-howto-tenant/)
```
Comment (saragluna) [inline sdk/spring/azure-spring-boot-samples/azure-spring-boot-sample-active-directory-spring-security-resource-server/README.md:22]: Change the link to reference style

**Review by saragluna — CHANGES_REQUESTED** [review]
Overall is great, but there are some minor changes that need to be addressed.

**Review by saragluna — APPROVED** [review]
LGTM
